### PR TITLE
[TACHYON-596] Improve hash of blockId so mapping to locks is more evenly distributed

### PR DESCRIPTION
**Pull request by ooq**
https://tachyon.atlassian.net/browse/TACHYON-596

### CODE REVIEW

File: servers/src/main/java/tachyon/worker/block/BlockLockManager.java [hunk 1]
```diff
@@ -70,8 +70,8 @@ public BlockLockManager(BlockMetadataManager metaManager) {
    * @throws IOException
    */
   public long lockBlock(long userId, long blockId, BlockLockType blockLockType) throws IOException {
-    // TODO: generate real hashValue on blockID.
-    int hashValue = (int) (blockId % (long) NUM_LOCKS);
+    // Knuth's multiplicative method: multiple by a number same order of hash size
+    int hashValue = (int) (Integer.toString((int) blockId).hashCode() * 997 % (long) NUM_LOCKS);
```
Comment (calvinjia) [inline servers/src/main/java/tachyon/worker/block/BlockLockManager.java]: I don't think you need to cast `NUM_LOCKS` to a long?

Comment (apc999) [inline servers/src/main/java/tachyon/worker/block/BlockLockManager.java]: check hash functions in guava here? 
http://docs.guava-libraries.googlecode.com/git-history/master/javadoc/com/google/common/hash/Hashing.html
I assume we don't need heavy-weight hash functions but choices like CRC might be sufficient.

**Comment by calvinjia** [comment]
@ooq Looks like this causes some of the tests to fail, could you take a look?

**Comment by calvinjia** [comment]
+1 LGTM

**Comment by apc999** [comment]
I think `consistentHash` is not the best choice---it provides properties that we don't need (i.e., incremental growth as bucket number gets more),  at O(n) space and O(logn) time complexity where n is the number of buckets. Why not use other hash functions provided by Guava? e.g. `goodFastHash` is a general-purpose, non-cryptographic-strength, streaming hash function that produces hash codes of length at least minimumBits.

in constructor:
```java
mHashFunc = Hashing.goodFastHash(32);  // hash func producing at least 32 bits output.
```
In lock:
```java
int hashValue = mHashFunc.hashLong(blockId).asInt() % NUM_LOCKS;
```

**Comment by apc999** [comment]
@ooq , by checking the implementation you pointed, it turns out the space cost is constant but the time complexity is still O(logn).

**Comment by ooq** [comment]
I skipped goodFastHash() because it uses system time as the seed. Not a fan of such un-determinism. Makes debugging harder. 
Also because I know very little of the math behind hashing, I'm not sure whether doing a modulo after hashing will weaken the randomness property of the result...
Is log(100) bad?

**Comment by apc999** [comment]
@ooq , `NUM_LOCKS` is 100 for now, but it is very likely we will increase its value to much bigger value (e.g., 1000 or much larger) to reduce hash collision.

I see your concern about non-deterministic hashing for `goodFastHash` and I agree with you. Thanks for pointing this out! How about using `crc32()` then? In our case, we don't need the hash to be strong but non-cryptographic and as cheap as possible.

**Comment by aarondav** [comment]
The tricky bit about CRC32 is that while it is extremely fast if it actually ends up using the x86 instruction, the lower bits are unsuitable for use as a hash function because incrementing sequences such as "1, 2, 3, ..." actually may end up having identical lower bits. That said, particular implementations can differ in which end has this property, so it requires a bit of experimentation. The "other end", though, makes for a very good hash function if you only need 16 bits. (See [this page](https://web.archive.org/web/20131120005827/http://home.comcast.net/~bretm/hash/8.html) for more info.)

For the sake of simplicity, though, I think either `Hashing.murmur3_32` (which is what Guava is using when you request <= 32 bits of "good, fast hashiness") or just consistentHash is fine, because it's likely that the volatile read to check the subsequent lock takes ten times as long as either of those.

**Comment by apc999** [comment]
@aarondav `murmur3_32` sounds good to me. A little bit more argument on `consistentHash`: the guava implementation of consistent hashing doesn't quite ensure uniformness across buckets --- people use virtual nodes to address this well-known imbalance problem but guava actually doesn't have it.

**Comment by ooq** [comment]
Thanks for the comments @apc999 @aarondav , this is a very educative discussion. Taken all the considerations above, I will go with `murmur3_32`.

**Comment by apc999** [comment]
@ooq  cool, thanks! Also you may want to open another PR to merge to `master` and close this one. In that PR, you should reference this PR.

**Comment by ooq** [comment]
The new PR is #1038.

**Pull request closed without merging**